### PR TITLE
Add global run timeout (`--run-timeout`)

### LINKED ISSUE #560: Global run timeout
Add a maximum wall-clock duration for the entire test run. If the total run exceeds this limit, karva would stop executing new tests and terminate.

```
karva test --run-timeout 30m
```

Or in config:

```toml
[test]
run-timeout = "30m"
```

This is a safety net for CI to prevent runaway test suites from consuming resources indefinitely.

See: https://nexte.st/docs/features/slow-tests/#configuring-timeouts

### PR DESCRIPTION
**Pull request by BitWeaverDev**
## Summary

This adds a wall-clock limit for the entire test run, which closes #560. It gives CI a safety net against runaway suites: if the whole run takes longer than the configured limit, karva stops the remaining workers and exits with a failure status instead of letting the job hang until the CI provider's own timeout kills it.

It can be set on the command line or in configuration:

```bash
karva test --run-timeout 1800
```

```toml
[profile.default.test]
run-timeout = 1800.0
```

I went with plain seconds (an `f64`) rather than the `"30m"` duration string sketched in the issue, to stay consistent with the existing `--timeout` and `--slow-timeout` flags and to avoid pulling in a duration-parsing dependency. Happy to switch to a humantime-style string if you'd prefer the issue's original shape.

The important design point is that this is **not** the same machinery as the per-test `timeout`. A per-test timeout is enforced inside the worker (in Python), so it gets forwarded to every worker process. The run timeout is a property of the run as a whole, so it lives entirely in the main process. The flag is therefore on `TestCommand` rather than `SubTestCommand`, and it is never forwarded to workers. Enforcement is a third early-exit condition in the existing `wait_for_completion` poll loop, sitting right next to the Ctrl+C shutdown and fail-fast checks that were already there. When the deadline passes, the loop returns a small `RunCompletion::TimedOut` outcome, the existing `kill_remaining` logic reaps the surviving workers, and the result is surfaced up as `RunOutput.timed_out`.

One edge case worth calling out: when a run is killed with nothing completed, the aggregated results are empty, which would otherwise look identical to "no tests were found". To avoid that confusion, the timeout is reported before the no-tests check, so the user sees `error: run timed out before all tests completed` and a non-zero exit rather than a misleading "no tests to run".

The configuration and CLI reference docs were regenerated with `cargo run -p karva_dev generate-all`.

## Test Plan

Added integration tests in `crates/karva/tests/it/run_timeout.rs` covering a run that exceeds the limit on the command line, the same via `karva.toml`, and a fast run that finishes comfortably under a generous limit (to confirm the flag doesn't disturb normal runs). The metadata unit-test snapshots were updated for the new option, and the existing `deny_unknown_fields` coverage now exercises the `run-timeout` key.

`just test` passes (728 integration tests, plus the workspace unit tests), and `uvx prek run -a` is clean aside from a pre-existing `ambiguous_derive_helpers` clippy warning in `karva_macros` that is unrelated to this change.

---

*I used an AI assistant to help with this contribution.*

### CODE REVIEW

File: crates/karva_cli/src/test.rs [hunk 1]
```diff
@@ -348,6 +356,7 @@ impl SubTestCommand {
                 no_tests: self.no_tests.map(Into::into),
                 slow_timeout: self.slow_timeout.map(SlowTimeoutSecs),
                 timeout: self.timeout.map(TestTimeoutSecs),
+                run_timeout: None,
```
Comment (MatthewMckee4) [inline crates/karva_cli/src/test.rs:380]: Why is this `None`?

**Comment by MatthewMckee4** [comment]
Thank you for this!

File: crates/karva_cli/src/test.rs (hunk 1 above)
Comment (BitWeaverDev) [reply crates/karva_cli/src/test.rs:380]: It's `None` here because `run-timeout` is a main process only option. It gets enforced in the orchestrator's worker wait loop and is never forwarded to workers, so I put the flag on `TestCommand` rather than on `SubTestCommand`, which is the struct shared with the worker binary.

Since `SubTestCommand` has no `run_timeout` field, it has nothing to supply at this point, so it leaves it `None`. The real value is filled in right afterwards in `TestCommand::into_options`:

```rust
let mut options = sub_command.into_options();
if let Some(test) = options.test.as_mut() {
    test.run_timeout = run_timeout.map(RunTimeoutSecs);
}
```

I kept it off `SubTestCommand` on purpose so the worker CLI doesn't pick up a `--run-timeout` flag that it would just ignore. If you'd rather it read more clearly, I'm happy to pull the `TestOptions` construction into a small constructor instead of the inline struct literal.

File: crates/karva_runner/src/orchestration.rs [hunk 2]
```diff
@@ -90,6 +105,13 @@ impl WorkerManager {
                 break;
             }
 
+            if let Some(timeout) = run_timeout
```
Comment (MatthewMckee4) [inline crates/karva_runner/src/orchestration.rs]: This checks the timeout before polling worker exits. If a worker finishes just before the deadline but the parent wakes after it, we can report a timeout even though all workers completed within the limit. Can we poll/reap workers first, return `Finished` if none remain, then check the timeout?

File: crates/karva_runner/src/orchestration.rs [hunk 3]
```diff
@@ -56,22 +65,28 @@ impl WorkerManager {
...
-            return;
+            return RunCompletion::Finished;
         }
 
         tracing::info!(
             "Waiting for {} workers to complete (Ctrl+C to cancel)",
             self.workers.len()
         );
 
+        let start = Instant::now();
```
Comment (MatthewMckee4) [inline crates/karva_runner/src/orchestration.rs]: In some other docs we say that the run timeout is for the whole run, but this is for just after collection, let's start the timer before collection, as early as possible, in this crate.

**Review by MatthewMckee4 — COMMENTED** [review]
Thank you.

We should also support this in watch mode

File: crates/karva_cli/src/test.rs (hunk 1 above)
Comment (MatthewMckee4) [reply crates/karva_cli/src/test.rs:380]: makes sense, thanks for the explanation

Comment (MatthewMckee4) [reply crates/karva_cli/src/test.rs:380]: Could we add a short comment above this line to explain this

**Comment by MatthewMckee4** [comment]
Just rebased there for you, since I pushed some quite major changes

File: crates/karva_runner/src/orchestration.rs (hunk 2 above)
Comment (BitWeaverDev) [reply crates/karva_runner/src/orchestration.rs]: Good catch. Fixed in 858599d. The loop now reaps finished workers at the top of each iteration and returns `AllCompleted` if none remain, and only then checks the shutdown, fail-fast, and timeout conditions. So a worker that exits just before the deadline is counted as completed instead of triggering a false timeout. I applied the same reap first ordering to the Ctrl+C and fail-fast checks since they had the same race.

File: crates/karva_runner/src/orchestration.rs (hunk 3 above)
Comment (BitWeaverDev) [reply crates/karva_runner/src/orchestration.rs]: Done in 858599d. The deadline is now computed at the very start of `run_parallel_tests`, before collection, and passed into `wait_for_completion` as an absolute `Instant`. So the limit covers the whole run now, which matches the docs. I also updated the doc comment on `wait_for_completion` to say the same.

File: crates/karva_cli/src/test.rs (hunk 1 above)
Comment (BitWeaverDev) [reply crates/karva_cli/src/test.rs:380]: Added in 858599d, a short comment above the line explaining that run-timeout is main process only and never forwarded to workers, so it is filled in by `TestCommand::into_options`.

**Comment by BitWeaverDev** [comment]
Pushed 858599d addressing the review:

* Reap finished workers before checking the shutdown, fail-fast, and timeout conditions, so a run that finishes right as the deadline passes is reported as completed rather than timed out.
* Anchor the deadline before collection so the limit covers the whole run.
* Report the timeout in watch mode too.
* Added the comment explaining why run_timeout is None in `SubTestCommand::into_options`.

Full suite is green (777 integration tests, 370 unit tests) and `prek` passes.

**Comment by MatthewMckee4** [comment]
Thank you very much!
